File: src/Billing/Api/BillingServiceImpl.cs
using Billing.Service.Api.Internal;
using Grpc.Core;
using Telcobright.Billing.V1;

namespace Billing.Service.Api;

/// <summary>
/// gRPC entry point for the rating / CDR / summary service — a THIN surface. Each RPC delegates to its
/// handler in <c>Billing.Service.Api.Internal</c>; there is NO logic here. These three methods ARE the
/// intent of the service:
/// <list type="bullet">
/// <item><see cref="ProcessCdrBatch"/> — rate + write a batch of cdrs for one tenant.</item>
/// <item><see cref="GetMaxRatePerMinute"/> — pre-call: quote the max rate per tier of the tenant chain.</item>
/// <item><see cref="FinalizeAndSummarize"/> — post-call: settle the call per level.</item>
/// </list>
/// </summary>
public sealed class BillingServiceImpl : RatingService.RatingServiceBase
{
    private readonly ProcessCdrBatchHandler _processCdrBatch;
    private readonly MaxRateHandler _maxRate;
    private readonly FinalizeHandler _finalize;

    public BillingServiceImpl(ProcessCdrBatchHandler processCdrBatch, MaxRateHandler maxRate, FinalizeHandler finalize)
    {
        _processCdrBatch = processCdrBatch;
        _maxRate = maxRate;
        _finalize = finalize;
    }

    public override Task<CdrBatchResult> ProcessCdrBatch(CdrBatchRequest request, ServerCallContext context)
        => _processCdrBatch.Handle(request, context);

    public override Task<MaxRateReply> GetMaxRatePerMinute(MaxRateRequest request, ServerCallContext context)
        => _maxRate.Handle(request, context);

    public override Task<FinalizeResponse> FinalizeAndSummarize(FinalizeRequest request, ServerCallContext context)
        => _finalize.Handle(request, context);
}

File: java/src/main/java/com/telcobright/billing/api/BillingServiceImpl.java
package com.telcobright.billing.api;

import com.telcobright.billing.api.internal.FinalizeHandler;
import com.telcobright.billing.api.internal.MaxRateHandler;
import com.telcobright.billing.api.internal.ProcessCdrBatchHandler;
import com.telcobright.billing.grpc.CdrBatchRequest;
import com.telcobright.billing.grpc.CdrBatchResult;
import com.telcobright.billing.grpc.FinalizeRequest;
import com.telcobright.billing.grpc.FinalizeResponse;
import com.telcobright.billing.grpc.MaxRateReply;
import com.telcobright.billing.grpc.MaxRateRequest;
import com.telcobright.billing.grpc.RatingService;
import io.quarkus.grpc.GrpcService;
import io.smallrye.mutiny.Uni;
import io.smallrye.mutiny.infrastructure.Infrastructure;
import jakarta.inject.Inject;

/**
 * The gRPC surface — a thin wrapper. Each RPC delegates to one internal handler; all logic lives there (and
 * in the beans/engine beneath). The blocking work runs on the worker pool so the gRPC event loop is never
 * blocked. Port of the .NET {@code BillingServiceImpl} (which delegated to the same three handlers).
 */
@GrpcService
public class BillingServiceImpl implements RatingService {

    private final ProcessCdrBatchHandler _processCdrBatch;
    private final MaxRateHandler _maxRate;
    private final FinalizeHandler _finalize;

    @Inject
    public BillingServiceImpl(ProcessCdrBatchHandler processCdrBatch, MaxRateHandler maxRate, FinalizeHandler finalize) {
        this._processCdrBatch = processCdrBatch;
        this._maxRate = maxRate;
        this._finalize = finalize;
    }

    @Override
    public Uni<MaxRateReply> getMaxRatePerMinute(MaxRateRequest request) {
        return Uni.createFrom().item(() -> _maxRate.Handle(request))
                .runSubscriptionOn(Infrastructure.getDefaultWorkerPool());
    }

    @Override
    public Uni<FinalizeResponse> finalizeAndSummarize(FinalizeRequest request) {
        return Uni.createFrom().item(() -> _finalize.Handle(request))
                .runSubscriptionOn(Infrastructure.getDefaultWorkerPool());
    }

    @Override
    public Uni<CdrBatchResult> processCdrBatch(CdrBatchRequest request) {
        return Uni.createFrom().item(() -> _processCdrBatch.Handle(request))
                .runSubscriptionOn(Infrastructure.getDefaultWorkerPool());
    }
}
